### PR TITLE
Adding density material to PorousFlow, closes #9027

### LINKED ISSUE #9027: Density used in Gravity kernel should vary with porosity and fluid density
### Description of the enhancement or error report
Add new Rock Density material to PorousFlow to enable density to vary with porosity and fluid density

### Rationale for the enhancement or information for reproducing the error
Rock density varies with porosity and fluid density, and this influences the body forces represented by the Gravity kernel

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)
New material added to PorousFlow module

### PR DESCRIPTION
**Pull request by hsheldon**
Added density material to PorousFlow that enables density varying with porosity
#9027

### CODE REVIEW

**Review by permcody — DISMISSED** [review]

**Comment by permcody** [comment]
@WilkAndy - This looks fine to me. I'll let you have a look over it though before it gets merged.

File: modules/porous_flow/include/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.h
Comment (WilkAndy) [inline modules/porous_flow/include/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.h]: I think change to

```
virtual Real rho_f() const;
```
only.  I can't foresee ever needing the phase number as an input.  If i'm wrong then it's easily changed, but i think making this general is unnecessary presently.  You may disagree!

File: modules/porous_flow/include/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.h
Comment (WilkAndy) [inline modules/porous_flow/include/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.h]: Remove the
```
unsigned int ph = 0
```
argument, and remove the default from v, ie just make it
```
unsigned int v
```
There's no need for the default.

File: modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C
Comment (WilkAndy) [inline modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C]: I'm not sure if you might want
```
rho_s >= 0
```
here instead.  I know that in reality rho_s must be positive, but i've been caught in the past when people have suddenly wanted to do some comparison with analytical solutions where some quantity has been ignored so they've wanted to set it zero, and my range check has prevented them.  Probably i'm being overly cautious here - i'll let you decide.

File: modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C
Comment (WilkAndy) [inline modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C]: If you follow my suggestion above, this line will just be
```
return _rho_f_qp[_qp][0];
```
in your case, and it'll be ready to be overridden by someone else.

File: modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C
Comment (WilkAndy) [inline modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C]: Probably you should check
```
_num_phases == 1
```
to prevent people from doing something too silly.

File: modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C
Comment (WilkAndy) [inline modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C]: and this line will be
```
return _drho_f_qp_dvar[_qp][0][v];
```

File: modules/porous_flow/tests/density/GravDensity01.i
Comment (WilkAndy) [inline modules/porous_flow/tests/density/GravDensity01.i]: block=0 is probably unnecessary

File: modules/porous_flow/tests/density/GravDensity01.i
Comment (WilkAndy) [inline modules/porous_flow/tests/density/GravDensity01.i]: No [Functions]

File: modules/porous_flow/tests/density/GravDensity01.i
Comment (WilkAndy) [inline modules/porous_flow/tests/density/GravDensity01.i]: almost all these Executioner parameters are unneeded

File: modules/porous_flow/tests/density/GravDensity01.i
Comment (WilkAndy) [inline modules/porous_flow/tests/density/GravDensity01.i]: it would be good to check the derivatives too.  i know that no kernel is actually using them, but it'd still be good.  Use a MaterialStdVectorAux

File: modules/porous_flow/tests/density/tests
Comment (WilkAndy) [inline modules/porous_flow/tests/density/tests]: probably don't need the rel_err parameter

File: modules/porous_flow/tests/density/GravDensity01.i
Comment (WilkAndy) [inline modules/porous_flow/tests/density/GravDensity01.i]: In which case, having a lower bulk density will be a good idea.  And maybe a more complicated porosity function - i'll let you decide as this could get really complicated quite quickly!

**Review by WilkAndy — CHANGES_REQUESTED** [review]
A few minor changes

File: modules/porous_flow/include/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.h
Comment (hsheldon) [inline modules/porous_flow/include/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.h]: ok

File: modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C
Comment (hsheldon) [inline modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C]: Good point, I did wonder about that.

File: modules/porous_flow/include/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.h
Comment (WilkAndy) [inline modules/porous_flow/include/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.h]: Oh, i just realised this function and drho_f_dvar should really go into the base class (TotalGravitationalDensityBase) as virtual pure functions if you want to override them in multiple classes.  Ie, in the TotalGravitationDensityBase.h you'd have
```
virtual Real rho_f() const = 0;
virtual Real drho_f_dvar(unsigned int v) const = 0;
```
and then you'd override as usual in your "FullySaturated" version.  **However** now i'm wondering whether the "override-these-functions" really is a wise design.  I think we're being unnecessarily complicated, and i think just remove these rho_f and drho_f_dvar functions altogether, and just have 

```
_gravdensity[_qp] = _rho_s * (1.0 - _porosity_qp[_qp]) + _rho_f_qp[_qp][0] * _porosity_qp[_qp];
```
in your computeQpProperties.  It's simpler, which is usually better.  If someone really wants to do some fancy overriding in the future then let them and don't imagine you know what they'll really want.  I'm sorry, i know i suggested these functions in the beginning, but i think i was wrong.

File: modules/porous_flow/tests/density/GravDensity01.i
Comment (hsheldon) [inline modules/porous_flow/tests/density/GravDensity01.i]: Hmm yes. The derivatives should be zero as it stands at the moment, because porosity is constant. So yes I would have to put in a more complicated porosity function to make the  derivative check meaningful.

File: modules/porous_flow/tests/density/GravDensity01.i
Comment (WilkAndy) [inline modules/porous_flow/tests/density/GravDensity01.i]: Probably you have to "git add" the changed csv file too

Comment (hsheldon) [inline modules/porous_flow/tests/density/GravDensity01.i]: Yes.... I shouldn't have pushed yet, still work in progress.

File: modules/porous_flow/tests/density/GravDensity01.i
Comment (hsheldon) [inline modules/porous_flow/tests/density/GravDensity01.i]: As discussed, you are correct there will be a non-zero derivative wrt fluid density. I will implement a check for that.

File: modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C
Comment (hsheldon) [inline modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C]: Done

File: modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C
Comment (hsheldon) [inline modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C]: Done

File: modules/porous_flow/tests/density/GravDensity01.i
Comment (hsheldon) [inline modules/porous_flow/tests/density/GravDensity01.i]: All good now, I hope!

File: modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C
Comment (WilkAndy) [inline modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C]: Excellent mooseError message!

**Review by WilkAndy — APPROVED** [review]
Looks good now.  Thanks @hsheldon.  As far as i'm concerned this can be merged if it passes the tests.

**Review by permcody — APPROVED** [review]

File: modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C
Comment (hsheldon) [inline modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C]: I think you wrote that originally, I copied it from elsewhere in PorousFlow!

**Pull request merged**
